handin-server/main.rkt

```diff
@@ -649,9 +650,12 @@
                 (proc (lambda ()
                         ;; Proc has succeeded...
                         (parameterize ([current-custodian orig-custodian])
-                          (kill-thread watcher))))
+                          (kill-thread watcher)
+                          ;; unblock the other putting thread if we killed the
+                          ;; watcher too fast.
+                          (channel-try-get watcher-channel))))
```

Comment:
It seems like this could be a problem, still, if the `channel-try-get` happens before the `channel-put` has started. Maybe this should just be `channel-get`, and the thread will get GCed when the channel becomes otherwise unreachable?

Reply:
Ah probably a good point. Just a second.

Reply:
Hm, when I do that the connection hangs when the client is not fast. I don't think the thread can be GCed, since the channel is shared among several threads, so it will just block.

Reply:
On further reflection, just in case `proc` retains a refetence to the thread in a way that means it can't be GCed, maybe it would be simpler and better to not use `channel-get` here and instead use `(sync (channel-put-evt watcher-channel session-thread) watcher)` below?